base/hosts.c

```diff
@@ -1749,6 +1807,193 @@ gvm_host_in_hosts (const gvm_host_t *host, const struct in6_addr *addr,
+                break;
+
+              if (host_type == HOST_TYPE_CIDR_BLOCK)
+                ips_func = cidr_block_ips;
+              else if (host_type == HOST_TYPE_RANGE_SHORT)
+                ips_func = short_range_network_ips;
+              else
+                ips_func = long_range_network_ips;
+
+              if (ips_func (stripped, &first, &last) == -1)
+                break;
+
+              /* Make sure that first actually comes before last */
+              if (ntohl (first.s_addr) > ntohl (last.s_addr))
+                break;
+
+              /* Check addresses */
+              if (ntohl (first.s_addr) <= ntohl (host->addr.s_addr)
+                  && ntohl (last.s_addr) >= ntohl (host->addr.s_addr))
+                found = 1;
+              break;
+            }
+          case HOST_TYPE_CIDR6_BLOCK:
+          case HOST_TYPE_RANGE6_LONG:
+          case HOST_TYPE_RANGE6_SHORT:
+            {
+              struct in6_addr first, last;
+              int (*ips_func) (const char *,
+                               struct in6_addr *, struct in6_addr *);
+
+              if (host->type != HOST_TYPE_IPV6)
+                break;
+
+              if (host_type == HOST_TYPE_CIDR6_BLOCK)
+                ips_func = cidr6_block_ips;
+              else if (host_type == HOST_TYPE_RANGE6_SHORT)
+                ips_func = short_range6_network_ips;
+              else
+                ips_func = long_range6_network_ips;
+
+              if (ips_func (stripped, &first, &last) == -1)
+                break;
+
+              /* Make sure the first comes before the last. */
+              if (memcmp (&first.s6_addr, &last.s6_addr, 16) > 0)
+                break;
+
+              /* Check addresses */
+              if (memcmp (&first.s6_addr,
+                          &(host->addr6).s6_addr, 16) <= 0
+                  && memcmp (&last.s6_addr,
+                             &host->addr6.s6_addr, 16) >= 0)
+                found = 1;
+
+              break;
+            }
+          case -1:
+          default:
+            {
+              /* Invalid host string. */
+              g_strfreev (split);
+              return FALSE;
+            }
+        }
+
+      host_element++;
+    }
+  g_strfreev (split);
+  return found;
+}
+
+/**
+ * @brief Returns whether a host has an equal host in a hosts string.
+ * eg. 192.168.10.1 has an equal in a hosts string
+ * "192.168.10.1-5, 192.168.10.10-20" string while 192.168.10.7 doesn't.
+ *
+ * @param[in] hosts_str       Hosts string to check.
+ * @param[in] find_host_str   The host string to find.
+ *
+ * @return 1 if host has equal in hosts_str, 0 otherwise.
+ */
+int
+gvm_hosts_str_contains (const char *hosts_str, const char *find_host_str)
+{
+  int ret;
+  gvm_host_t *host;
+
+  host = gvm_host_from_str (find_host_str);
+
+  if (host == NULL)
+    return 0;
+
+  ret = gvm_host_in_hosts_str (host, hosts_str);
+  gvm_host_free (host);
+
+  return ret;
+}
+
```

Comment:
Actually, the whole patch could be discarded by using something like this in GVM:

{
gvm_hosts_t *hosts = gvm_hosts_new (hosts_str);
gvm_hosts_t *hosts2 = gvm_hosts_new (find_host_str);
if (!hosts || !hosts2 || hosts2->count != 0)
{
gvm_hosts_free (hosts);
gvm_hosts_free (hosts2);
return 0
}

int ret = gvm_host_in_hosts (hosts2->hosts->data, NULL, hosts);
gvm_hosts_free (hosts);
gvm_hosts_free (hosts2);
return ret;
}

Reply:
I wanted to avoid larger allocations that would occur for ranges and CIDR blocks since the function could be called for a large number of results.

Reply:
Hi @timopollmeier I did some more tests (with tests/test-hosts) to see where the hosts module is i.r.w memory allocations and speed. In order to get a really big memory footprint / speed impact, one has to use large host numbers (ie. millions of hosts) which would have the same impact everywhere else (scans etc,.)

I believe that adding 200LOC that pretty much duplicates existing functionality instead of using a couple lines shouldn't be done, unless there is a really good reason for it.

Reply:
@kroosec, I've updated the manager PR to use the existing libraries function like you suggested, with one change: I've used gvm_hosts_new_with_max(...) to avoid issues with large allocation in case a large hosts range is passed to the function.

Reply:
I'll be closing this PR now since it's no longer necessary with the new solution for GVM.